src/machine/eval.rs

```diff
@@ -17,12 +17,15 @@ impl<W: Write, A: Arithmetic> Machine<W, A> {
             BinaryOp(op, t1, t2) => self.step_bin_op(op, t1, t2),
             // Dispatch step for unary operations
             UnaryOp(op, t1) => self.step_un_op(op, t1),
-            // Dispatch step for beta reduction
-            // Application with unevaluated first term (t1 t2)
-            // Evaluate t1.
-            App(mut t1, arg) => match *t1 {
+            App(mut t1, mut arg) => match *t1 {
+                // Dispatch step for beta reduction
                 Abs(body) => self.step_beta_reduction(*body, arg),
-                PrimFn(prim) => self.step_primitive_app(prim, arg),
+                // Dispatch step for primitive application if argument is evaluated
+                PrimFn(prim) if arg.is_value() => self.step_primitive_app(prim, arg),
+                // Evaluate argument if that's not the case.
+                PrimFn(_) => (self.step_in_place(arg.borrow_mut()), App(t1, arg)),
```

Comment:
Can the evaluation of `arg` be moved inside of `step_primitive_app` so we could alter the behavior depending on the primitive?

Reply:
I'd require wrapping the `Primitive` in a `Term::PrimFn` again which would be slower.

I've been thinking 💭 about a way to fork execution and then go back to where we left to avoid all those `step_in_place`. But there is some groundwork we'd need to do first

Reply:
Or we take the whole handling of `PrimFn` in its own function to the required matching. Or we add a helper function like `evaluate_arg() -> bool` to the `Primitive`s and evaluate it depending on the result. This is only necessary if we want to be able to handle unevaluated args in `Primitive`s.

Reply:
Now that I think about it we do the unwrapping and rewrapping for `Term::Cond` inside `step_conditional` because we need the condition to continue evaluation. I agree, we need some sort of helper mechanism for this.

Reply:
I added (in https://github.com/christianpoveda/pijama/pull/107/commits/38b8d6b970a7ad1cb9387264125fdd9c53f95c3d) a new function `eval` which tries to evaluate a term repeatedly until it is not possible to do anything else (this is basically big-step evaluation). This gives us the nice advantage of just evaluating a term we know we need to keep going.

Here are some benchmarks:
```
arithmetic              time:   [5.9519 us 6.0662 us 6.2686 us]                        
                        change: [-72.604% -72.081% -71.554%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  2 (2.00%) high mild
  5 (5.00%) high severe

logic                   time:   [6.7219 us 6.7626 us 6.8114 us]                   
                        change: [-84.712% -84.574% -84.438%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 13 outliers among 100 measurements (13.00%)
  5 (5.00%) high mild
  8 (8.00%) high severe

factorial               time:   [34.276 us 34.338 us 34.411 us]                       
                        change: [-41.268% -40.748% -40.212%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 9 outliers among 100 measurements (9.00%)
  1 (1.00%) high mild
  8 (8.00%) high severe

factorial_tail          time:   [93.371 us 95.047 us 97.862 us]                           
                        change: [-10.396% -9.1458% -7.4607%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 11 outliers among 100 measurements (11.00%)
  4 (4.00%) high mild
  7 (7.00%) high severe

fibonacci               time:   [357.07 us 359.91 us 363.70 us]                      
                        change: [-24.544% -23.031% -21.277%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 17 outliers among 100 measurements (17.00%)
  4 (4.00%) high mild
  13 (13.00%) high severe

fibonacci_tail          time:   [87.783 us 88.406 us 89.128 us]                           
                        change: [-6.6899% -5.1332% -3.4851%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 15 outliers among 100 measurements (15.00%)
  1 (1.00%) low mild
  8 (8.00%) high mild
  6 (6.00%) high severe

gcd                     time:   [556.13 us 557.64 us 559.41 us]                
                        change: [-23.620% -22.648% -21.851%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 15 outliers among 100 measurements (15.00%)
  2 (2.00%) low mild
  3 (3.00%) high mild
  10 (10.00%) high severe

ackermann               time:   [358.76 us 366.43 us 376.74 us]                      
                        change: [-2.3636% -0.6120% +1.5221%] (p = 0.52 > 0.05)
                        No change in performance detected.
Found 15 outliers among 100 measurements (15.00%)
  3 (3.00%) high mild
  12 (12.00%) high severe

calling                 time:   [147.66 us 150.82 us 155.58 us]                    
                        change: [-5.8342% -3.5243% -1.4286%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 10 outliers among 100 measurements (10.00%)
  4 (4.00%) high mild
  6 (6.00%) high severe

complex_calling         time:   [154.52 us 156.02 us 157.94 us]                            
                        change: [-12.197% -8.5473% -5.5234%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 12 outliers among 100 measurements (12.00%)
  5 (5.00%) high mild
  7 (7.00%) high severe

fancy_max               time:   [1.5366 us 1.5575 us 1.5817 us]                       
                        change: [-7.0247% -5.7207% -4.4612%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 12 outliers among 100 measurements (12.00%)
  1 (1.00%) low mild
  4 (4.00%) high mild
  7 (7.00%) high severe

step                    time:   [1.2148 us 1.2224 us 1.2317 us]                  
                        change: [-3.6915% -1.5866% +0.9787%] (p = 0.20 > 0.05)
                        No change in performance detected.
Found 15 outliers among 100 measurements (15.00%)
  1 (1.00%) high mild
  14 (14.00%) high severe
```

Reply:
That are some good improvements in terms of speed and removing `step_in_place` is also a win because it was a complicated function. 🎆